File: packages/editor/src/beat-map.ts
import type { ChordEvent, NoteEvent, SyncAnchor, TabDocument } from './types'

export interface BeatMap {
  tempo_bpm: number
  time_signature: [number, number]
  bar_offset_ticks: number
  sync_anchors: SyncAnchor[]
}

export type BeatMapValidationCode =
  | 'anchor-count'
  | 'anchor-order'
  | 'boundary'
  | 'meter'
  | 'offset'
  | 'tempo'
  | 'segment-tempo'

export class BeatMapValidationError extends Error {
  readonly code: BeatMapValidationCode
  readonly anchorIndex?: number

  constructor(code: BeatMapValidationCode, message: string, anchorIndex?: number) {
    super(message)
    this.name = 'BeatMapValidationError'
    this.code = code
    this.anchorIndex = anchorIndex
  }
}

export interface TempoEvent {
  scoreTick: number
  bpm: number
}

export interface BeatMapCoverage {
  startFrame: number
  endFrame: number
}

type TabWithBeatMap = TabDocument & { bar_offset_ticks?: number }

export function ticksPerMeasure(
  ticksPerQuarter: number,
  timeSignature: readonly [number, number],
): number {
  const [beats, beatType] = timeSignature
  if (
    !Number.isInteger(beats) ||
    beats <= 0 ||
    beats > 32 ||
    !Number.isInteger(beatType) ||
    ![2, 4, 8, 16].includes(beatType)
  ) {
    throw new BeatMapValidationError(
      'meter',
      'Time signature needs 1–32 beats and a beat unit of 2, 4, 8, or 16.',
    )
  }
  const ticks = (beats * ticksPerQuarter * 4) / beatType
  if (!Number.isInteger(ticks)) {
    throw new BeatMapValidationError(
      'meter',
      'Time signature cannot use whole score ticks at this resolution.',
    )
  }
  return ticks
}

export function beatMapFromTab(tab: TabDocument): BeatMap {
  return {
    tempo_bpm: tab.tempo_bpm,
    time_signature: [...tab.time_signature] as [number, number],
    bar_offset_ticks: (tab as TabWithBeatMap).bar_offset_ticks ?? 0,
    sync_anchors: tab.sync_anchors.map((anchor) => ({ ...anchor })),
  }
}

function inferredCoverage(tab: TabDocument): BeatMapCoverage | undefined {
  const events = [...tab.notes, ...tab.chords.events]
  if (!events.length) return undefined
  return {
    startFrame: Math.min(...events.map((event) => event.onset_frame)),
    endFrame: Math.max(...events.map((event) => event.end_frame)),
  }
}

export function validateBeatMap(
  tab: TabDocument,
  beatMap: BeatMap,
  coverage: BeatMapCoverage | undefined = inferredCoverage(tab),
): BeatMap {
  if (!Number.isFinite(beatMap.tempo_bpm) || beatMap.tempo_bpm <= 20 || beatMap.tempo_bpm > 400) {
    throw new BeatMapValidationError('tempo', 'Tempo must be greater than 20 and at most 400 BPM.')
  }
  const measureTicks = ticksPerMeasure(tab.ticks_per_quarter, beatMap.time_signature)
  if (
    !Number.isInteger(beatMap.bar_offset_ticks) ||
    beatMap.bar_offset_ticks < 0 ||
    beatMap.bar_offset_ticks >= measureTicks
  ) {
    throw new BeatMapValidationError(
      'offset',
      `Pickup must be between 0 and ${measureTicks - 1} ticks.`,
    )
  }
  if (beatMap.sync_anchors.length < 2 || beatMap.sync_anchors.length > 5000) {
    throw new BeatMapValidationError('anchor-count', 'Beat Map needs 2–5000 sync pins.')
  }
  beatMap.sync_anchors.forEach((anchor, index) => {
    if (
      !Number.isSafeInteger(anchor.audio_frame) ||
      anchor.audio_frame < 0 ||
      !Number.isSafeInteger(anchor.score_tick) ||
      anchor.score_tick < 0
    ) {
      throw new BeatMapValidationError(
        'anchor-order',
        'Sync pins need non-negative whole frames and score ticks.',
        index,
      )
    }
    if (index === 0) return
    const previous = beatMap.sync_anchors[index - 1]
    if (
      anchor.audio_frame <= previous.audio_frame ||
      anchor.score_tick <= previous.score_tick
    ) {
      throw new BeatMapValidationError(
        'anchor-order',
        'Sync pins must move forward in both audio and score time.',
        index,
      )
    }
    const frameSpan = anchor.audio_frame - previous.audio_frame
    const tickSpan = anchor.score_tick - previous.score_tick
    const bpm = (tickSpan * tab.sample_rate * 60) / (frameSpan * tab.ticks_per_quarter)
    if (!Number.isFinite(bpm) || bpm <= 20 || bpm > 400) {
      throw new BeatMapValidationError(
        'segment-tempo',
        `Pins ${index} and ${index + 1} imply ${bpm.toFixed(1)} BPM; use 20–400 BPM.`,
        index,
      )
    }
  })
  if (
    coverage &&
    (beatMap.sync_anchors[0].audio_frame > coverage.startFrame ||
      beatMap.sync_anchors.at(-1)!.audio_frame < coverage.endFrame)
  ) {
    throw new BeatMapValidationError(
      'boundary',
      'Locked first and last sync pins must cover the transcription passage.',
    )
  }

  return {
    tempo_bpm: beatMap.tempo_bpm,
    time_signature: [...beatMap.time_signature] as [number, number],
    bar_offset_ticks: beatMap.bar_offset_ticks,
    sync_anchors: beatMap.sync_anchors.map((anchor) => ({ ...anchor })),
  }
}

/** Piecewise-linear map. Extrapolation uses nearest segment for boundary events. */
export function scoreTickForAudioFrame(audioFrame: number, anchors: readonly SyncAnchor[]): number {
  if (anchors.length < 2) throw new BeatMapValidationError('anchor-count', 'Beat Map needs at least 2 sync pins.')
  let left = anchors[0]
  let right = anchors[1]
  if (audioFrame >= anchors.at(-1)!.audio_frame) {
    left = anchors.at(-2)!
    right = anchors.at(-1)!
  } else {
    for (let index = 1; index < anchors.length; index += 1) {
      if (audioFrame <= anchors[index].audio_frame) {
        left = anchors[index - 1]
        right = anchors[index]
        break
      }
    }
  }
  const progress = (audioFrame - left.audio_frame) / (right.audio_frame - left.audio_frame)
  return Math.max(0, Math.round(left.score_tick + progress * (right.score_tick - left.score_tick)))
}

function remapEvent<T extends NoteEvent | ChordEvent>(event: T, anchors: SyncAnchor[]): T {
  const scoreTick = scoreTickForAudioFrame(event.onset_frame, anchors)
  const endTick = scoreTickForAudioFrame(event.end_frame, anchors)
  return {
    ...event,
    score_tick: scoreTick,
    duration_ticks: Math.max(1, endTick - scoreTick),
  }
}

/**
 * Replace timing as one immutable operation. Audio frame/second fields are copied unchanged;
 * score fields are regenerated from the staged anchors.
 */
export function applyBeatMap(
  tab: TabDocument,
  candidate: BeatMap,
  coverage?: BeatMapCoverage,
): TabDocument & { bar_offset_ticks: number } {
  const beatMap = validateBeatMap(tab, candidate, coverage ?? inferredCoverage(tab))
  return {
    ...tab,
    tempo_bpm: beatMap.tempo_bpm,
    time_signature: beatMap.time_signature,
    bar_offset_ticks: beatMap.bar_offset_ticks,
    sync_anchors: beatMap.sync_anchors,
    notes: tab.notes.map((note) => remapEvent(note, beatMap.sync_anchors)),
    chords: {
      ...tab.chords,
      events: tab.chords.events.map((chord) => remapEvent(chord, beatMap.sync_anchors)),
    },
  }
}

/** Tempo changes implied by adjacent pins, suitable for MIDI and MusicXML export. */
export function tempoEventsForTab(tab: TabDocument): TempoEvent[] {
  const events: TempoEvent[] = [{ scoreTick: 0, bpm: tab.tempo_bpm }]
  const anchors = tab.sync_anchors
  for (let index = 0; index + 1 < anchors.length; index += 1) {
    const left = anchors[index]
    const right = anchors[index + 1]
    const bpm =
      ((right.score_tick - left.score_tick) * tab.sample_rate * 60) /
      ((right.audio_frame - left.audio_frame) * tab.ticks_per_quarter)
    if (!Number.isFinite(bpm) || bpm <= 20 || bpm > 400) continue
    const rounded = Math.round(bpm * 1000) / 1000
    const previous = events.at(-1)!
    if (left.score_tick === previous.scoreTick) {
      previous.bpm = rounded
    } else if (Math.abs(previous.bpm - rounded) >= 0.001) {
      events.push({ scoreTick: left.score_tick, bpm: rounded })
    }
  }
  return events
}

export function pickupMidiShift(tab: TabDocument): number {
  const offset = beatMapFromTab(tab).bar_offset_ticks
  if (offset === 0) return 0
  return ticksPerMeasure(tab.ticks_per_quarter, tab.time_signature) - offset
}

File: server/solotrace/beat_map.py
from __future__ import annotations

from dataclasses import dataclass
from math import isfinite

from pydantic import Field, field_validator, model_validator

from .models import ChordEvent, NoteEvent, StrictModel, SyncAnchor, TabDocument


class BeatMap(StrictModel):
    tempo_bpm: float = Field(gt=20, le=400)
    time_signature: tuple[int, int] = (4, 4)
    bar_offset_ticks: int = Field(default=0, ge=0)
    sync_anchors: list[SyncAnchor] = Field(min_length=2, max_length=5000)

    @field_validator("time_signature")
    @classmethod
    def validate_time_signature(cls, value: tuple[int, int]) -> tuple[int, int]:
        beats, beat_type = value
        if not 1 <= beats <= 32 or beat_type not in {2, 4, 8, 16}:
            raise ValueError(
                "Time signature needs 1–32 beats and a beat unit of 2, 4, 8, or 16"
            )
        return value

    @model_validator(mode="after")
    def validate_anchor_order(self) -> BeatMap:
        for left, right in zip(self.sync_anchors, self.sync_anchors[1:], strict=False):
            if right.audio_frame <= left.audio_frame or right.score_tick <= left.score_tick:
                raise ValueError(
                    "Sync pins must move forward in both audio and score time"
                )
        return self


@dataclass(frozen=True)
class TempoEvent:
    score_tick: int
    bpm: float


def ticks_per_measure(
    ticks_per_quarter: int,
    time_signature: tuple[int, int],
) -> int:
    beats, beat_type = time_signature
    numerator = beats * ticks_per_quarter * 4
    if numerator % beat_type:
        raise ValueError(
            "Time signature cannot use whole score ticks at this resolution"
        )
    return numerator // beat_type


def beat_map_from_tab(tab: TabDocument) -> BeatMap:
    anchors = tab.sync_anchors
    if len(anchors) < 2:
        # Existing projects can contain one pin. A staged edit still needs locked
        # start/end pins, so callers must establish them before applying.
        raise ValueError("Beat Map needs at least 2 sync pins")
    return BeatMap(
        tempo_bpm=tab.tempo_bpm,
        time_signature=tab.time_signature,
        bar_offset_ticks=getattr(tab, "bar_offset_ticks", 0),
        sync_anchors=anchors,
    )


def _inferred_coverage(tab: TabDocument) -> tuple[int, int] | None:
    events = [*tab.notes, *tab.chords.events]
    if not events:
        return None
    return (
        min(event.onset_frame for event in events),
        max(event.end_frame for event in events),
    )


def validate_beat_map(
    tab: TabDocument,
    beat_map: BeatMap,
    coverage: tuple[int, int] | None = None,
) -> BeatMap:
    measure_ticks = ticks_per_measure(tab.ticks_per_quarter, beat_map.time_signature)
    if beat_map.bar_offset_ticks >= measure_ticks:
        raise ValueError(f"Pickup must be between 0 and {measure_ticks - 1} ticks")
    required_coverage = coverage or _inferred_coverage(tab)
    if required_coverage is not None:
        start_frame, end_frame = required_coverage
        if (
            beat_map.sync_anchors[0].audio_frame > start_frame
            or beat_map.sync_anchors[-1].audio_frame < end_frame
        ):
            raise ValueError(
                "Locked first and last sync pins must cover the transcription passage"
            )
    for index, (left, right) in enumerate(
        zip(beat_map.sync_anchors, beat_map.sync_anchors[1:], strict=False),
        start=1,
    ):
        frame_span = right.audio_frame - left.audio_frame
        tick_span = right.score_tick - left.score_tick
        bpm = tick_span * tab.sample_rate * 60 / (frame_span * tab.ticks_per_quarter)
        if not isfinite(bpm) or not 20 < bpm <= 400:
            raise ValueError(
                f"Pins {index} and {index + 1} imply {bpm:.1f} BPM; use 20–400 BPM"
            )
    return beat_map


def score_tick_for_audio_frame(audio_frame: int, anchors: list[SyncAnchor]) -> int:
    if len(anchors) < 2:
        raise ValueError("Beat Map needs at least 2 sync pins")
    left, right = anchors[0], anchors[1]
    if audio_frame >= anchors[-1].audio_frame:
        left, right = anchors[-2], anchors[-1]
    else:
        for candidate_left, candidate_right in zip(anchors, anchors[1:], strict=False):
            if audio_frame <= candidate_right.audio_frame:
                left, right = candidate_left, candidate_right
                break
    progress = (audio_frame - left.audio_frame) / (right.audio_frame - left.audio_frame)
    return max(0, round(left.score_tick + progress * (right.score_tick - left.score_tick)))


def _remap_event(
    event: NoteEvent | ChordEvent,
    anchors: list[SyncAnchor],
) -> NoteEvent | ChordEvent:
    score_tick = score_tick_for_audio_frame(event.onset_frame, anchors)
    end_tick = score_tick_for_audio_frame(event.end_frame, anchors)
    return event.model_copy(
        update={
            "score_tick": score_tick,
            "duration_ticks": max(1, end_tick - score_tick),
        }
    )


def apply_beat_map(
    tab: TabDocument,
    candidate: BeatMap,
    coverage: tuple[int, int] | None = None,
) -> TabDocument:
    """Replace score timing while leaving source audio coordinates untouched."""
    beat_map = validate_beat_map(tab, candidate, coverage)
    notes = [
        _remap_event(note, beat_map.sync_anchors)
        for note in tab.notes
    ]
    chords = tab.chords.model_copy(
        update={
            "events": [
                _remap_event(chord, beat_map.sync_anchors)
                for chord in tab.chords.events
            ]
        }
    )
    return tab.model_copy(
        update={
            "tempo_bpm": beat_map.tempo_bpm,
            "time_signature": beat_map.time_signature,
            "bar_offset_ticks": beat_map.bar_offset_ticks,
            "sync_anchors": beat_map.sync_anchors,
            "notes": notes,
            "chords": chords,
        }
    )


def tempo_events_for_tab(tab: TabDocument) -> list[TempoEvent]:
    events = [TempoEvent(score_tick=0, bpm=tab.tempo_bpm)]
    for left, right in zip(tab.sync_anchors, tab.sync_anchors[1:], strict=False):
        bpm = (
            (right.score_tick - left.score_tick)
            * tab.sample_rate
            * 60
            / ((right.audio_frame - left.audio_frame) * tab.ticks_per_quarter)
        )
        if not isfinite(bpm) or not 20 < bpm <= 400:
            continue
        rounded = round(bpm, 3)
        if left.score_tick == events[-1].score_tick:
            events[-1] = TempoEvent(score_tick=left.score_tick, bpm=rounded)
        elif abs(events[-1].bpm - rounded) >= 0.001:
            events.append(TempoEvent(score_tick=left.score_tick, bpm=rounded))
    return events


def pickup_midi_shift(tab: TabDocument) -> int:
    offset = getattr(tab, "bar_offset_ticks", 0)
    if not offset:
        return 0
    return ticks_per_measure(tab.ticks_per_quarter, tab.time_signature) - offset
